pkg/controller/dnszone/dnszone_controller_test.go

```diff
@@ -506,6 +506,88 @@ func TestReconcileDNSProviderForAzure(t *testing.T) {
 	}
 }
 
+// TestReconcileDNSProviderForAWSWithConditions tests that expected conditions are set after calling ReconcileDNSProvider for AWS
+func TestReconcileDNSProviderForAWSWithConditions(t *testing.T) {
+	log.SetLevel(log.DebugLevel)
+
+	cases := []struct {
+		name               string
+		dnsZone            *hivev1.DNSZone
+		setupAWSMock       func(*awsmock.MockClientMockRecorder)
+		errorExpected      bool
+		expectDnsCondition bool
+		dnsCondition       hivev1.DNSZoneCondition
+		actuator           string
+	}{
+		{
+			name:    "Fail to create hosted zone, set generic dns error condition",
+			dnsZone: validDNSZoneWithoutID(),
+			setupAWSMock: func(expect *awsmock.MockClientMockRecorder) {
+				mockAWSZoneDoesntExist(expect, validDNSZoneWithoutID())
+				mockAWSZoneCreationError(expect)
+			},
+			errorExpected:      true,
+			expectDnsCondition: true,
+			dnsCondition: hivev1.DNSZoneCondition{
+				Type:    hivev1.GenericDNSErrorsCondition,
+				Status:  corev1.ConditionTrue,
+				Reason:  dnsCloudErrorReason,
+				Message: "KMSOptInRequired: error creating hosted zone\ncaused by: error creating hosted zone",
+			},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Arrange
+			mocks := setupDefaultMocks(t)
+
+			zr, _ := NewAWSActuator(
+				log.WithField("controller", ControllerName),
+				mocks.fakeKubeClient,
+				awsclient.CredentialsSource{},
+				tc.dnsZone,
+				fakeAWSClientBuilder(mocks.mockAWSClient),
+			)
+
+			r := ReconcileDNSZone{
+				Client: mocks.fakeKubeClient,
+				logger: zr.logger,
+				scheme: scheme.Scheme,
+			}
+
+			// This is necessary for the mocks to report failures like methods not being called an expected number of times.
+			defer mocks.mockCtrl.Finish()
+
+			setFakeDNSZoneInKube(mocks, tc.dnsZone)
+
+			if tc.setupAWSMock != nil {
+				tc.setupAWSMock(mocks.mockAWSClient.EXPECT())
+			}
+
+			// Act
+			_, err := r.reconcileDNSProvider(zr, tc.dnsZone)
+			zr.SetConditionsForError(err)
```

Comment:
I might have taken the easier route here - implementing entire Reconcille Test was too complicated

Reply:
I think it's completely fair to break out a separate test, mega suites that do all kinds of complicated things are not a pattern I like.

However I am confused by these two lines. zr is an actuator, it's passed into the reconciler r, which we call, why would we explicitly call SetConditionsForError on the actuator after running reconcile, which returned the error?

Reply:
So there's no test in dnsZone controller that tests the Reconcile itself. There are tests for reconcileDNSProvider and SetConditionsForError. For this particular situation, I needed to call both methods because actuators return an error(if any) via reconcileDNSProvider, whereas SetConditionsForError parses the error and decides how the conditions would be set. It mimics [this](https://github.com/openshift/hive/blob/master/pkg/controller/dnszone/dnszone_controller.go#L216-L217) code block of Reconcile.